kubeportal/k8s/kubernetes_api.py

```diff
@@ -73,13 +129,182 @@ def get_service_accounts():
+            stripped_svc = {'name': svc.metadata.name,
+                            'type': svc.spec.type,
+                            'selector': svc.spec.selector,
+                            'creation_timestamp': svc.metadata.creation_timestamp}
+            ports = []
+            for port in svc.spec.ports:
+                ports.append({"port": port.port, "protocol": port.protocol})
+            stripped_svc["ports"] = ports
+            stripped_services.append(stripped_svc)
+        return stripped_services
+    except Exception as e:
+        logger.exception(f"Error while fetching services of namespace {namespace}")
+        return []
+
+
+def create_k8s_service(namespace: str, name: str, svc_type: str, selector: dict, ports: list):
+    """
+    Create a Kubernetes service in the cluster.
+    The 'ports' parameter contains a list of dictionaries, each with the key 'port' and 'protocol'.
+    """
+    logger.info(f"Creating Kubernetes service '{name}'")
+    svc_ports = [client.V1ServicePort(name=str(p["port"]), port=p["port"], protocol=p["protocol"]) for p in ports]
+
+    svc = V1Service(
+        metadata=client.V1ObjectMeta(name=name),
+        spec=V1ServiceSpec(
+            type=svc_type,
+            selector=selector,
+            ports=svc_ports
+        )
+    )
+    core_v1.create_namespaced_service(namespace, svc)
+
+
+def create_k8s_ingress(namespace: str, name: str, annotations: dict, tls: bool, rules: dict):
+    """
+    Create a Kubernetes ingress in the cluster.
+
+    The Kubeportal API makes some simplifying assumptions:
+
+    - TLS is a global configuration for an Ingress.
+    - The necessary annotations come from the portal settings, not from the ingress definition.
+    """
+    logger.info(f"Creating Kubernetes ingress '{name}'")
+
+    import pdb
+    pdb.set_trace()
+
+    k8s_ing = NetworkingV1beta1Ingress(
+              metadata=client.V1ObjectMeta(name=name, annotations=annotations)
+    )
+
+    k8s_rules = []
+    for host, host_config in rules.items():
+        k8s_rule = NetworkingV1beta1IngressRule(host=host)
+        k8s_paths = []
+        for path, path_config in host_config.items():
+            k8s_backend = NetworkingV1beta1IngressBackend(
+                service_name=path_config['service_name'],
+                service_port=path_config['service_port']
+            )
+            k8s_paths.append(NetworkingV1beta1HTTPIngressPath(path=path, backend=k8s_backend))
+        k8s_http = NetworkingV1beta1HTTPIngressRuleValue(
+            paths=k8s_paths
+        )
+        k8s_rule.http = k8s_http
+        k8s_rules.append(k8s_rule)
+
+    k8s_spec = NetworkingV1beta1IngressSpec(rules=k8s_rules)
+
+    if tls:
+        k8s_ing.metadata.annotations['cert-manager.io/cluster-issuer'] = settings.INGRESS_TLS_ISSUER
+        k8s_spec.tls = [NetworkingV1beta1IngressTLS(hosts=list(rules.keys()), secret_name=f'{name}_tls')]
+
+    k8s_ing.spec = k8s_spec
+    #TODO: networking.k8s.io/v1 Ingress
```

Comment:
what does it mean? create_k8s_ingress() with an old ingress api version shouldn't be possible....

Reply:
I had strange effects when running the KubePortal Ingress API against K8S 1.20 in Minikube. I need much better test coverage for ingress, so that CI/CD can check the KubePortal API against different K8S versions.